libraries/AP_OSD/AP_OSD_Backend.cpp

```diff
@@ -24,32 +24,63 @@ constexpr uint8_t AP_OSD_Backend::symbols[AP_OSD_NUM_SYMBOLS];
 #define SYM_DIG_OFS_1 0x90
 #define SYM_DIG_OFS_2 0xA0
 
+
+uint8_t AP_OSD_Backend::convert_to_decimal_packed_characters(char* buff, uint8_t size) {
+#if OSD_ENABLED
+    // use packed decimal characters based on fiam idea implemented in inav osd
+    // search the decimal separator with a bound and always terminate the string
+    char* p = (char*)memchr(&buff[1],'.',size-1);
+    if (p && isdigit(p[1]) && isdigit(p[-1])) {
+        // remove the decimal separator and replace the digit before and after
+        p[-1] += SYM_DIG_OFS_1;
+        p[1] += SYM_DIG_OFS_2;
+        // shift anything after p[1] 1 character to the left
+        const char* move_start = p+1;
+        const uint8_t move_size = size-(move_start-buff);
+        memmove(p, move_start, move_size);
+        p[move_size] = 0x00; // terminate
+        return size-1;
+    }
+#else
+    // we guarantee string is terminated
+    buff[size-1] = 0x00;
+#endif
+    return size;
+}
+
+uint8_t AP_OSD_Backend::format_string_for_osd(char* buff, uint8_t size, bool decimal_packed, const char *fmt, va_list ap)
+{
+#if OSD_ENABLED
+    // note: vsnprintf() always terminates the string
+    int res = hal.util->vsnprintf(buff, size, fmt, ap);
+    res = MIN(res, size);
+    if (res > 0 && decimal_packed) {
+        // note: convert_to_decimal_packed_characters() always terminates the string
+        res = convert_to_decimal_packed_characters(buff, res);
+    }
+    return res;
+#else
+    // we guarantee string is terminated
+    buff[0] = 0x00;
```

Comment:
Also need an explicit check for size==0 in this routine.

Reply:
yep, you're right once again!